packages/wdio-sauce-service/src/service.ts

```diff
@@ -166,52 +231,61 @@ export default class SauceService {
         }
 
         const status = 'status: ' + (failures > 0 ? 'failing' : 'passing')
-        if (!global.browser.isMultiremote) {
+        if (!this.multiRemoteBrowser.isMultiremote) {
             log.info(`Update job with sessionId ${global.browser.sessionId}, ${status}`)
             return this.isUP ? this.updateUP(failures) : this.updateJob(global.browser.sessionId, failures)
         }
 
-        return Promise.all(Object.keys(this.capabilities).map((browserName) => {
-            log.info(`Update multiremote job for browser "${browserName}" and sessionId ${global.browser[browserName].sessionId}, ${status}`)
-            return this.isUP ? this.updateUP(failures) : this.updateJob(global.browser[browserName].sessionId, failures, false, browserName)
-        }))
+        if (this.capabilities) {
+            return Promise.all(Object.keys(this.capabilities).map((browserName) => {
+                const browser = global.browser[browserName as keyof BrowserObject] as BrowserObject
+                log.info(`Update multiremote job for browser "${browserName}" and sessionId ${browser.sessionId}, ${status}`)
+                return this.isUP ? this.updateUP(failures) : this.updateJob(browser.sessionId, failures, false, browserName)
+            }))
+        }
     }
 
-    onReload (oldSessionId, newSessionId) {
+    onReload (oldSessionId: string, newSessionId: string) {
         if (!this.isServiceEnabled && !this.isRDC) {
             return
         }
 
         const status = 'status: ' + (this.failures > 0 ? 'failing' : 'passing')
 
-        if (!global.browser.isMultiremote) {
+        if (!this.multiRemoteBrowser.isMultiremote) {
             log.info(`Update (reloaded) job with sessionId ${oldSessionId}, ${status}`)
             return this.updateJob(oldSessionId, this.failures, true)
         }
 
-        const browserName = global.browser.instances.filter(
-            (browserName) => global.browser[browserName].sessionId === newSessionId)[0]
+        const browserName = this.multiRemoteBrowser.instances.filter(
+            (browserName: string) => (global.browser[browserName as keyof BrowserObject] as BrowserObject).sessionId === newSessionId)[0]
```

Comment:
I would say this as quite hacky @christian-bromann. `browserName` is definitely not in the list of `keyof BrowserObject`

Reply:
Since the type of `global.browser` is `BrowserObject`, and you are trying to access a key from `global.browser` here, what else could it be than a key of `BrowserObject`?

Reply:
I think this is the case of accessing something like `Record<string, BrowserObject>`. If you use `<keyOf>`, it would be only declared properties of `BrowserObject`, so the typecheck would be meaning less, because `browserName` is not in the declared properties.

Reply:
> If you use <keyOf>, the property will be casted to any, so the typecheck would provide no value.

That's actually incorrect, have a look at the following example:
```ts
const myObj = {
  Hello: "world"
};
type ObjKey = keyof typeof myObj
const foo: ObjKey = "foo";
```
If what you are saying was correct, that `keyof` would cast it to `any`, then I would expect this example to compile without any errors (since `ObjKey` would be `any`). However, this results in the following error:
```
Initializer type "foo" is not assignable to variable type ObjKey
```
If I change the last line to be `const foo: ObjKey = "Hello";`, it compiles fine without any errors.
So clearly TypeScript preserves the types when using `keyof`.

But you are right in a way that by using `browserName as keyof BrowserObject` we are implicitly telling TypeScript to _assume_ that `browserName` is a key of `BrowserObject`, even though it comes from `this.multiRemoteBrowser.instances` (which is a plain array of strings, with no guarantee of those strings being `keyof BrowserObject`). To be 100% correct we would need to either type `instance` as `(keyof BrowserObject)[]` and or run checks on runtime to ensure that `instances` are indeed keys of `BrowserObject`. But this would limit the instances that can be used, which could get annoying to maintain (as the code needs to be changed for each new type of instance) and it would most likely be overkill here and the risk of breaking things would be much higher than the benefit of the type safety we would gain in this situation, in my opinion.

Considering the code has been running fine previously, I'm quite hesitant with adding any more checks here, we could always consider improving type safety here in the future, but probably this would be a bigger undertaking. Of course I could be totally missing something, in which case I would be glad if you could point it out or make suggestions on how to improve the situation.